### DISCUSSION TITLE
Address Search Functionality

### DISCUSSION THREAD

**Discussion by biljohns** [janoside/btc-rpc-explorer#440] [Q&A] [answered]
I was curious, how did this project manage to implement a search by address function in the explorer navbar?

I am running a bitcoin core node and have the bitcoin-cli functioning. However, there is no command to give the cli an address and return a balance. Well, there is, but you have to set the address as "watched". I am trying to return address balances on the fly (as with the search function in this application).

I see many files with "address" in the file names at the following location, `btc-rpc-explorer/app/api/`.
_Do I have to use something else with bitcoin core in order to be able to search by address to get the balances?_

Any information or help would be greatly appreciated.

**Comment by jsarenik** [comment]
The address search is set in `BTC_ADDRESS_API`. See the related part of [.env-sample](https://github.com/janoside/btc-rpc-explorer/blob/87fe724a357e29fef0dd433b6d065a2ac55a3aab/.env-sample#L41-L48).

**Comment by jsarenik** [comment]
One can run their local instance of a minimalistic https://github.com/bwt-dev/bwt, but then the search bar will not be able to search for any address, just those monitored by `bwt` (usually defined by an `xpub`-like).

**Comment by jsarenik** [comment, marked as answer]
No matter what, even without any `BTC_ADDRESS_API`, the search for a Transaction ID, the first-class-citizen of Bitcoin, should succeed. Then what the Bitcoin Core's wallet is doing is to parse all the newcoming blocks and see if any of the transactions refers the address we have the matching private key of.

Indexing all the addresses and what transactions they were used in is quite expensive/data-hungry (see Blockstream's electrs indexing requirements](https://github.com/Blockstream/electrs#installing--indexing). The original [electrs](https://github.com/romanz/electrs) seems to have much smaller requirements.

**Reply by biljohns** [reply]
So, the missing piece for me would be the electrs server I would have to have running in conjunction with bitcoin core?
Does electrs index the bitcoin blockchain into a database that is queryable?

**Reply by jsarenik** [reply]
The missing piece could also be using some public Electrum server, or as you say, run electrs yourself. Yes, use https://electrumx-spesmilo.readthedocs.io/en/latest/protocol.html as a queryable API.
